Housekeeping for 0.7

jamwaffles opened the pull request:
Progress towards #8 and #5, closes #7, closes #6.

jamwaffles commented:
Rectangle intersection moved to embedded-graphics/debug-tools#9.

jamwaffles commented:
A bit of a kitchen sink PR, but I think I've covered most points in the other issues in this repo

rfuest commented on eg-next/examples/primitives-fill.rs:
```diff
@@ -24,7 +24,7 @@ where
         .draw(target)?;
 
-    Rectangle::new(Point::new(96, 0), Size::new(64, 64))
+    Rectangle::new(Point::new(97, 0), Size::new(64, 64))
```
Why is this change necessary? (Maybe related to the comment below)

jamwaffles replied:
Well spotted, yes that was to fix some clipping. I've now fixed the new clipping this changed caused so we should be all good now.

rfuest commented on eg-next/examples/primitives-fill.rs:
```diff
@@ -33,21 +33,21 @@ where
         .into_styled(style)
         .draw(target)?;
```
I think the `stroke_off_fill_of` rectangle clips the lower left corner of the `stroke` triangle.

rfuest commented on eg-next/examples/primitives-fill.rs:
```diff
@@ -33,21 +33,21 @@ where
 
 fn main() -> Result<(), Infallible> {
-    let mut display: SimulatorDisplay<BinaryColor> = SimulatorDisplay::new(Size::new(512, 128));
+    let mut display: SimulatorDisplay<BinaryColor> = SimulatorDisplay::new(Size::new(300, 200));
```
Could we move all drawn shapes a bit to the bottom right, so that they look centered? Either by adding a `translated` call to all `draw_shapes` invocations or by adding an offset inside `draw_shaped`.

jamwaffles replied:
Done. I've also added some spacing too. The previous `BinaryColor` rendering was actually pretty centered, just the invisible final shape was moving the "visual center" over a bit.

rfuest commented on eg-next/examples/primitives-fill.rs:
```diff
@@ -33,21 +33,21 @@ where
+    let mut display: SimulatorDisplay<BinaryColor> = SimulatorDisplay::new(Size::new(300, 200));
 
     let stroke = PrimitiveStyle::with_stroke(BinaryColor::On, 1);
 
```
The `stroke_off_fill_off` style behaves the same as a `PrimitiveStyle::with_fill(BinaryColor::Off)`. Would it be better to change the example to use colors to better differentiate the styles?

rfuest commented on eg-next/examples/primitives-triangles.rs:
```diff
@@ -1,60 +1,96 @@
+    // Inside thick stroke, no fill
+    base_triangle
+        .into_styled(
+            PrimitiveStyleBuilder::new()
+                .stroke_alignment(StrokeAlignment::Inside)
+                .stroke_width(10)
+                .stroke_color(Rgb888::CSS_SALMON)
```
I try to follow this order in the examples:
```suggestion
            PrimitiveStyleBuilder::new()
                .stroke_color(Rgb888::CSS_SALMON)
                .stroke_width(10)
                .stroke_alignment(StrokeAlignment::Inside)
```
The reason to use this order is that the first two calls are in the same order as the arguments in `PrimitiveStyle::with_stroke`. And the alignment is an advanced option that should follow the basic settings IMO.

rfuest commented on eg-next/examples/primitives-triangles.rs:
```diff
@@ -1,60 +1,96 @@
+        .translate(Point::new(offset, 0))
+        .into_styled(
+            PrimitiveStyleBuilder::new()
+                .stroke_width(10)
+                .stroke_color(Rgb888::CSS_AQUAMARINE)
+                .fill_color(Rgb888::CSS_CADET_BLUE)
```
```suggestion
                .stroke_color(Rgb888::CSS_AQUAMARINE)
                .stroke_width(10)
                .fill_color(Rgb888::CSS_CADET_BLUE)
```

rfuest commented on eg-next/examples/primitives-triangles.rs:
```diff
@@ -1,60 +1,96 @@
+        .translate(Point::new(offset * 2, 0))
+        .into_styled(
+            PrimitiveStyleBuilder::new()
+                .stroke_alignment(StrokeAlignment::Outside)
+                .stroke_width(9)
+                .stroke_color(Rgb888::CSS_FIRE_BRICK)
+                .fill_color(Rgb888::CSS_WHITE_SMOKE)
```
```suggestion
                .stroke_color(Rgb888::CSS_FIRE_BRICK)
                .stroke_width(9)
                .stroke_alignment(StrokeAlignment::Outside)
                .fill_color(Rgb888::CSS_WHITE_SMOKE)
```

rfuest commented on eg-next/examples/text-alignment.rs:
```diff
@@ -0,0 +1,60 @@
+        "Left aligned text, origin top left",
+        bounding_box.top_left,
+        character_style,
+        TextStyleBuilder::new()
+            .alignment(Alignment::Left)
+            .baseline(Baseline::Top)
+            .build(),
```
I would suggest to define all styles before drawing any text.

This is a general suggestions for all inline style builder calls. I've seen multiple examples where people have written wrappers around text drawing or other draw functions because our calls seem to be to complicated to use every time. But IMO many of these problems could have been solved by creating styles ahead of time.

jamwaffles replied:
I think I've caught every instance of `TextStyleBuilder` and `PrimitiveStyleBuilder` and moved them to their own variables. Let me know if I missed one.

rfuest commented on eg-next/examples/text-alignment.rs:
```diff
@@ -0,0 +1,60 @@
+    .draw(&mut display)?;
+
+    let output_settings = OutputSettingsBuilder::new().scale(2).build();
+    Window::new("Fonts", &output_settings).show_static(&display);
```
```suggestion
    Window::new("Text alignment", &output_settings).show_static(&display);
```

rfuest commented on eg-next/examples/text-styles.rs:
```diff
@@ -0,0 +1,71 @@
+    Text::new(" text!", position, normal).draw(&mut display)?;
+
+    let output_settings = OutputSettingsBuilder::new().scale(2).build();
+    Window::new("Fonts", &output_settings).show_static(&display);
```
Nice example!

```suggestion
    Window::new("Text styles", &output_settings).show_static(&display);
```

rfuest reviewed (changes requested):
First round of reviews

rfuest commented:
> Closes #8, closes #7, closes #6, closes #5

#8 still need some work after this PR is done.

And this PR is a good start towards #5, but could also use some work. Should we also add licensing information to the readme?

rfuest commented:
* Should we move `simulator-input-handling` to the examples directory in the `simulator` repo?
* Should we rename `line-thickness` to `primitives-line-thickness` or also move it `debug-tools`?
* Can we set the initial progress value in `demo-progress` to something != 0? This would make the screenshot more interesting.
* `text-custom-font` will need to be fixed after the next beta release.

jamwaffles commented:
To address your points in order:

1. Agreed, that example really doesn't make sense here does it. Moved in embedded-graphics/simulator#26
2. I've deleted it, as we already have the `line` debug tool which does the same thing AFAICS.
3. Set to >0 now, looks better indeed.
4. :+1: 

As for the licenses, I can use the same structure as the e-g repo. Let me do that quick...

jamwaffles commented:
Licenses are added, MIT or Apache 2.0 the same as e-g. I also added them to the `Cargo.toml` of each folder which isn't really necessary, but can't hurt.

rfuest commented:
Some final comments:
* I would prefer to keep the analog prefix for the clock demo, because I'm considering to add a `eg-seven-segment` based digital clock example to this repo: https://github.com/embedded-graphics/eg-seven-segment/blob/master/examples/digital-clock.rs
* IMO `primitives-ellipse` and `primitives-rounded-rectangles` are debug tools. Should we remove them? Both primitives are used in other examples.
* The comment in `text-fonts` says: "Demonstrate the available builtin fonts." We should either change the comment to state that this is a subset of available fonts and a full list is included in the docs or add all fonts to the example. I guess the first solution would be easier to implement and maintain in the future.
* `text-extended-characters`: The comment needs to be updated, because all builtin fonts support the characters in this example

jamwaffles commented:
Changes addressed :)

rfuest commented on LICENSE-MIT:
```diff
@@ -0,0 +1,25 @@
+Copyright (c) 2020 James Waples
+
```
Other creates in the Rust ecosystem (including the main rust repo) have removed this line from the MIT license. I think the TLDR was that each author retains the copyright to their contribution and that a global copyright is invalid anyway.

```suggestion
```

jamwaffles replied:
Makes sense. I've removed it from the license file.

rfuest reviewed (changes requested):
Looks good, only one small issue left.

rfuest commented:
Should we "fork" the eg-0.7 examples directory in this PR? Because the `Cargo.toml` in `eg-next` was supposed to always use git dependencies.

jamwaffles commented:
Ah, no, I think we should do that when 0.7 is released. Technically there's no proper release yet, so keeping `eg-next` as is for the short term makes sense to me. I've reverted the dependency changes back to the git/patches for this PR.

rfuest commented:
OK. But I have to admit that finally seeing the `Cargo.toml` without any git dependencies looked really good.

jamwaffles commented:
Thanks for the reviews! These examples look really good now.

> But I have to admit that finally seeing the Cargo.toml without any git dependencies looked really good.

It did look spectacular! The quicker we can get the migration guide done the quicker we can release non-beta and go back to non-Git `Cargo.toml` utopia ;)
